pcl_detector/src/pcl_detector_ros.cpp

```diff
@@ -0,0 +1,143 @@
+  
+  // Initialize the detector with configured parameters
+  std::string detector;
+  get_parameter<std::string>("detector", detector);
+  m_detector = initialize_detector(detector);
+}
+
+// Callback function for parameter changes
+rcl_interfaces::msg::SetParametersResult PclDetectorNode::parametersCallback(const std::vector<rclcpp::Parameter> &parameters) {
+    std::string detector_name;
+    rcl_interfaces::msg::SetParametersResult result;
+
+    for (const auto &parameter : parameters) {
+        if (parameter.get_name() == "detector") {
+            if (detector_type.count(parameter.as_string()) == 0) {
+                result.successful = false;
+                return result;
+            }
+        }
+    }
+    
+    parameters_changed = true;
+    result.successful = true;
+    result.reason = "success";
+    return result;
+}
+
+// Initialize the selected detector based on the parameter
+std::unique_ptr<IPclDetector> PclDetectorNode::initialize_detector(std::string detector)
+{
+    std::unique_ptr<IPclDetector> configured_detector = nullptr;
+
+    switch (detector_type[detector]) {
+        case DetectorType::Euclidean: {
+            float cluster_tolerance;
+            get_parameter<float>("euclidean.cluster_tolerance", cluster_tolerance);
+            int min_points;
+            get_parameter<int>("euclidean.min_points", min_points);
+            configured_detector = std::make_unique<EuclideanClusteringDetector>(cluster_tolerance, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "Euclidean detector initialized with cluster tolerance = " << cluster_tolerance << " and min_points = " << min_points);
+            break;
+        }
+
+        case DetectorType::DBSCAN: {
+            float eps; 
+            get_parameter<float>("dbscan.epsilon", eps);
+            int min_points;
+            get_parameter<int>("dbscan.min_points", min_points);
+            configured_detector = std::make_unique<DBSCANDetector>(eps, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "DBSCAN detector initialized with eps = " << eps << " and min_points = " << min_points);
+            break;
+        }
+
+        default: {
+            RCLCPP_ERROR(this->get_logger(), "Invalid pcl detector specified! Shutting down...");
+            rclcpp::shutdown();
+        }
+    }
+
+    return configured_detector;
+}
+
+// Callback function for processing incoming point cloud messages
+void PclDetectorNode::topic_callback(const sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg)
+{
+    // Check if parameters have changed
+    if (parameters_changed) {
+        std::string detector;
+        get_parameter<std::string>("detector", detector);
+        
+        m_detector = initialize_detector(detector);
+        parameters_changed = false;
+    }
+
+    // Converts incoming PointCloud2-msg to a PointCloud 
+    pcl::PointCloud<pcl::PointXYZ>::Ptr cartesian_cloud(new pcl::PointCloud<pcl::PointXYZ>);
```

Comment:
I really hope `::Ptr` is an alias for a `shared_ptr` or `unique_ptr` type because using the `new` keyword is a good recipe for memory leaks. It is probably fine, but you should look into doing something like 
```cpp
auto cartesian_cloud = std::make_shared<pcl::PointCloud<pcl::PointXYZ>();
```
instead.

Reply:
This is the preferred way from the PCL people, ::Ptr is an alias for boost::shared_ptr iirc, so this is fine. (I also remember something about using std::make_shared resulting in unnecessary deep copies of the cloud, which hurts performance quite a bit, but don't quote me on that)

NOTE however that `new` is a code smell, and is only acceptable in this case due to how the PCL is implemented.